src/highlighter.ts

```diff
@@ -0,0 +1,113 @@
+import { TextBuffer, LanguageMode } from "atom"
+
+declare module "atom" {
+  interface GrammarRegistry {
+    grammarForId(id: string): Grammar
+    languageModeForGrammarAndBuffer(g: Grammar, b: TextBuffer): LanguageMode
+  }
+  interface LanguageMode {
+    readonly fullyTokenized?: boolean
+    readonly tree?: boolean
+    onDidTokenize(cb: () => void): Disposable
+    buildHighlightIterator(): HighlightIterator
+    classNameForScopeId(id: ScopeId): string
+    startTokenizing?(): void
+  }
+  interface HighlightIterator {
+    seek(pos: { row: number; column: number }): void
+    getPosition(): { row: number; column: number }
+    getOpenScopeIds?(): ScopeId[]
+    getCloseScopeIds?(): ScopeId[]
+    moveToSuccessor(): void
+  }
+  interface ScopeId {}
+}
+
+export async function highlightTreeSitter(sourceCode: string, scopeName: string) {
+  const yielder = await eventLoopYielder(100, 5000)
+  const buf = new TextBuffer()
+  try {
+    const grammar = atom.grammars.grammarForId(scopeName)
+    const lm = atom.grammars.languageModeForGrammarAndBuffer(grammar, buf)
+    buf.setLanguageMode(lm)
+    buf.setText(sourceCode)
+    const end = buf.getEndPosition()
+    if (lm.startTokenizing) lm.startTokenizing()
+    await tokenized(lm)
+    const iter = lm.buildHighlightIterator()
+    if (iter.getOpenScopeIds && iter.getCloseScopeIds) {
+      let pos = { row: 0, column: 0 }
+      iter.seek(pos)
+      const res = []
+      while (pos.row < end.row || (pos.row === end.row && pos.column <= end.column)) {
+        const open = iter.getOpenScopeIds().map((x) => lm.classNameForScopeId(x))
+        const close = iter.getCloseScopeIds().map((x) => lm.classNameForScopeId(x))
+        res.push(...close.map((_) => `</span>`))
+        res.push(...open.map((x) => `<span class="${x}">`))
```

Comment:
Are these two the same?
```suggestion
        res.push(
          /** open */ ...iter.getOpenScopeIds().map((x) => `<span class="${lm.classNameForScopeId(x)}">`),
          /** close */ ...iter.getCloseScopeIds().map((_) => "</span>")
        )
```

Reply:
We first must push `close`, then `open`, due to how iterator is structured. Other than that, probably the same in terms of behaviour. In terms of performance, I would expect V8 should fuse map invocations, but it's fickle, so maybe not.

Reply:
So, the following works for sure
```suggestion
        // Note: don't inline! the iterators should be calculated in order
        const close = iter.getCloseScopeIds().map(() => "</span>")
        const open = iter.getOpenScopeIds().map((x) => `<span class="${lm.classNameForScopeId(x)}">`)
        res.push(...close, ...open)
```

Reply:
or is it just pushing that should happen in order? If that is the only requirement, then the first suggestion is fine because push always pushes from the start.

Reply:
No, no, the order of `get*ScopeIds()` invocations doesn't matter, only the actual order those are pushed to `res`. Besides, in JS arguments are evaluated left-to-right by spec.

Reply:
Right, so the first suggestion works.

Reply:
You've edited it after I started replying :shrug: